// element-plus/lib/components/breadcrumb/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var vue = require('vue');

var script$1 = vue.defineComponent({
  name: "ElBreadcrumb",
  props: {
    separator: {
      type: String,
      default: "/"
    },
    separatorClass: {
      type: String,
      default: ""
    }
  },
  setup(props) {
    const breadcrumb = vue.ref(null);
    vue.provide("breadcrumb", props);
    vue.onMounted(() => {
      const items = breadcrumb.value.querySelectorAll(".el-breadcrumb__item");
      if (items.length) {
        items[items.length - 1].setAttribute("aria-current", "page");
      }
    });
    return {
      breadcrumb
    };
  }
});

const _hoisted_1$1 = {
  ref: "breadcrumb",
  class: "el-breadcrumb",
  "aria-label": "Breadcrumb",
  role: "navigation"
};
function render$1(_ctx, _cache, $props, $setup, $data, $options) {
  return vue.openBlock(), vue.createElementBlock("div", _hoisted_1$1, [
    vue.renderSlot(_ctx.$slots, "default")
  ], 512);
}

script$1.render = render$1;
script$1.__file = "packages/components/breadcrumb/src/index.vue";

var script = vue.defineComponent({
  name: "ElBreadcrumbItem",
  props: {
    to: {
      type: [String, Object],
      default: ""
    },
    replace: {
      type: Boolean,
      default: false
    }
  },
  setup(props) {
    const link = vue.ref(null);
    const parent = vue.inject("breadcrumb");
    const instance = vue.getCurrentInstance();
    const router = instance.appContext.config.globalProperties.$router;
    vue.onMounted(() => {
      link.value.setAttribute("role", "link");
      link.value.addEventListener("click", () => {
        if (!props.to || !router)
          return;
        props.replace ? router.replace(props.to) : router.push(props.to);
      });
    });
    return {
      link,
      separator: parent == null ? void 0 : parent.separator,
      separatorClass: parent == null ? void 0 : parent.separatorClass
    };
  }
});

const _hoisted_1 = { class: "el-breadcrumb__item" };
const _hoisted_2 = {
  key: 1,
  class: "el-breadcrumb__separator",
  role: "presentation"
};
function render(_ctx, _cache, $props, $setup, $data, $options) {
  return vue.openBlock(), vue.createElementBlock("span", _hoisted_1, [
    vue.createElementVNode("span", {
      ref: "link",
      class: vue.normalizeClass(["el-breadcrumb__inner", _ctx.to ? "is-link" : ""]),
      role: "link"
    }, [
      vue.renderSlot(_ctx.$slots, "default")
    ], 2),
    _ctx.separatorClass ? (vue.openBlock(), vue.createElementBlock("i", {
      key: 0,
      class: vue.normalizeClass(["el-breadcrumb__separator", _ctx.separatorClass])
    }, null, 2)) : (vue.openBlock(), vue.createElementBlock("span", _hoisted_2, vue.toDisplayString(_ctx.separator), 1))
  ]);
}

script.render = render;
script.__file = "packages/components/breadcrumb/src/item.vue";

script$1.install = (app) => {
  app.component(script$1.name, script$1);
  app.component(script.name, script);
};
script$1.BreadcrumbItem = script;
const _Breadcrumb = script$1;
const ElBreadcrumb = _Breadcrumb;
const ElBreadcrumbItem = script;

exports.ElBreadcrumb = ElBreadcrumb;
exports.ElBreadcrumbItem = ElBreadcrumbItem;
exports['default'] = _Breadcrumb;
